src/main.cpp

```diff
@@ -1915,10 +1918,23 @@ void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, int nHeight)
 bool CScriptCheck::operator()() {
     const CScript &scriptSig = ptxTo->vin[nIn].scriptSig;
     const CScriptWitness *witness = (nIn < ptxTo->wit.vtxinwit.size()) ? &ptxTo->wit.vtxinwit[nIn].scriptWitness : NULL;
-    if (!VerifyScript(scriptSig, scriptPubKey, witness, nFlags, CachingTransactionSignatureChecker(ptxTo, nIn, amount, cacheStore), &error)) {
-        return false;
+    CachedHashes cachedHashes;
+    int version;
+    std::vector<unsigned char> program;
+    // if the cache map is available, and we get a cache miss, then we'll need to update the cache later
+    // if the scriptPubKey is not witness program, no need for cachedHashes, so do not read the map (it would acquire the lock needlessly)
+    bool addCachedHashesToMap = cachedHashesMap != NULL &&
+                                scriptPubKey.IsWitnessProgram(version, program) &&
+                                !cachedHashesMap->TryGet(this->ptxTo->GetHash(), &cachedHashes);
+    addCachedHashesToMap &= !cachedHashes.IsFull();
+    bool result = VerifyScript(scriptSig, scriptPubKey, witness, nFlags, CachingTransactionSignatureChecker(ptxTo, nIn, amount, cacheStore, cachedHashes), &error);
+    // do not update the cache if cachedHashes is still empty, it can happen if the evaluation ends before any segwit CheckSig operation could run
+    addCachedHashesToMap &= !cachedHashes.IsEmpty();
```

Comment:
Might make sense to check if any "progress" was made instead to avoid grabbing locks for setting any time the retrieved cachedHashes isn't empty already.